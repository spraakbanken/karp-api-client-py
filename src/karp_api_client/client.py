import os
import ssl
import typing
from typing import Any, Optional, Self, TypeVar, Union

import attrs
import httpx


class ApiKeyAuth(httpx.Auth):
    def __init__(self, token: str) -> None:
        self.api_token = token

    def auth_flow(
        self, request: httpx.Request
    ) -> typing.Generator[httpx.Request, httpx.Response, None]:
        request.url = request.url.copy_with(api_key=self.api_token)
        yield request


T = TypeVar("T", bound="ClientBase")


@attrs.define(slots=False)
class ClientBase:
    raise_on_unexpected_status: bool = attrs.field(default=False, kw_only=True)
    _base_url: str = attrs.field(
        default="https://spraakbanken4.it.gu.se/karp/v7", alias="base_url"
    )
    _cookies: dict[str, str] = attrs.field(factory=dict, kw_only=True, alias="cookies")
    _headers: dict[str, str] = attrs.field(factory=dict, kw_only=True, alias="headers")
    _timeout: Optional[httpx.Timeout] = attrs.field(
        default=None, kw_only=True, alias="timeout"
    )
    _verify_ssl: Union[str, bool, ssl.SSLContext] = attrs.field(
        default=True, kw_only=True, alias="verify_ssl"
    )
    _follow_redirects: bool = attrs.field(
        default=False, kw_only=True, alias="follow_redirects"
    )
    _httpx_args: dict[str, Any] = attrs.field(
        factory=dict, kw_only=True, alias="httpx_args"
    )
    _client: Optional[httpx.Client] = attrs.field(default=None, init=False)
    _async_client: Optional[httpx.AsyncClient] = attrs.field(default=None, init=False)

    def set_base_url(self, base_url: str) -> Self:
        """Update the base_url for this Client."""
        self._base_url = base_url
        return self

    def with_headers(self, headers: dict[str, str]) -> Self:
        """Get a new client matching this one with additional headers"""
        if self._client is not None:
            self._client.headers.update(headers)
        if self._async_client is not None:
            self._async_client.headers.update(headers)
        return attrs.evolve(self, headers={**self._headers, **headers})

    def with_cookies(self, cookies: dict[str, str]) -> Self:
        """Get a new client matching this one with additional cookies"""
        if self._client is not None:
            self._client.cookies.update(cookies)
        if self._async_client is not None:
            self._async_client.cookies.update(cookies)
        return attrs.evolve(self, cookies={**self._cookies, **cookies})

    def with_timeout(self, timeout: httpx.Timeout) -> Self:
        """Get a new client matching this one with a new timeout (in seconds)"""
        if self._client is not None:
            self._client.timeout = timeout
        if self._async_client is not None:
            self._async_client.timeout = timeout
        return attrs.evolve(self, timeout=timeout)

    def set_sync_client(self, client: httpx.Client) -> Self:
        """Manually set the underlying httpx.Client

        **NOTE**: This will override any other settings on the client, including cookies, headers, and timeout.
        """
        self._client = client
        return self

    def get_sync_client(self) -> httpx.Client:
        """Get the underlying httpx.Client, constructing a new one if not previously set"""
        if self._client is None:
            self._client = self._create_sync_client()
        return self._client

    # @abc.abstractmethod
    def _create_sync_client(self) -> httpx.Client:
        return httpx.Client(
            base_url=self._base_url,
            cookies=self._cookies,
            headers=self._headers,
            timeout=self._timeout,
            verify=self._verify_ssl,
            follow_redirects=self._follow_redirects,
            **self._httpx_args,
        )

    def __enter__(self) -> Self:
        """Enter a context manager for self.client—you cannot enter twice (see httpx docs)"""
        self.get_sync_client().__enter__()
        return self

    def __exit__(self, *args: Any, **kwargs: Any) -> None:
        """Exit a context manager for internal httpx.Client (see httpx docs)"""
        self.get_sync_client().__exit__(*args, **kwargs)

    def set_async_client(self, async_client: httpx.AsyncClient) -> Self:
        """Manually the underlying httpx.AsyncClient

        **NOTE**: This will override any other settings on the client, including cookies, headers, and timeout.
        """
        self._async_client = async_client
        return self

    def get_async_client(self) -> httpx.AsyncClient:
        """Get the underlying httpx.AsyncClient, constructing a new one if not previously set"""
        if self._async_client is None:
            self._async_client = self._create_async_client()
        return self._async_client

    def _create_async_client(self) -> httpx.AsyncClient:
        return httpx.AsyncClient(
            base_url=self._base_url,
            cookies=self._cookies,
            headers=self._headers,
            timeout=self._timeout,
            verify=self._verify_ssl,
            follow_redirects=self._follow_redirects,
            **self._httpx_args,
        )

    async def __aenter__(self) -> Self:
        """Enter a context manager for underlying httpx.AsyncClient—you cannot enter twice (see httpx docs)"""
        await self.get_async_client().__aenter__()
        return self

    async def __aexit__(self, *args: Any, **kwargs: Any) -> None:
        """Exit a context manager for underlying httpx.AsyncClient (see httpx docs)"""
        await self.get_async_client().__aexit__(*args, **kwargs)


@attrs.define(slots=False)
class Client(ClientBase):
    """Client to use for unauthenticated API calls."""


@attrs.define(slots=False)
class AuthenticatedClient(ClientBase):
    """Client to use for authenticated API calls."""

    _token: str = attrs.field(kw_only=True, alias="token")

    def _create_sync_client(self) -> httpx.Client:
        client = super()._create_sync_client()
        client.auth = ApiKeyAuth(self._token)
        return client

    def _create_async_client(self) -> httpx.AsyncClient:
        client = super()._create_async_client()
        client.auth = ApiKeyAuth(self._token)
        return client

    @classmethod
    def from_env(cls) -> "AuthenticatedClient":
        token = None
        if token_from_env := os.environ.get("KARP_API_CLIENT_API_TOKEN"):
            token = token_from_env
        elif token_from_env := os.environ.get("KARP_API_TOKEN"):
            token = token_from_env

        if token is None:
            raise RuntimeError("must set KARP_API_CLIENT_API_TOKEN or KARP_API_TOKEN")
        else:
            return cls(token=token)
