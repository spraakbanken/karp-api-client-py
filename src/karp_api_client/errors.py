"""Contains shared errors types that can be raised from API functions."""


class UnexpectedStatus(Exception):  # noqa: N818
    """Raised by api functions when the response status is an undocumented status.

    Only raised if Client.raise_on_unexpected_status is True.

    Args:
        status_code: the returned status code
        content: the content of the response
    """

    def __init__(self, status_code: int, content: bytes) -> None:
        self.status_code = status_code
        self.content = content

        super().__init__(
            f"Unexpected status code: {status_code}\n\nResponse content:\n{content.decode(errors='ignore')}"
        )


__all__ = ["UnexpectedStatus"]
