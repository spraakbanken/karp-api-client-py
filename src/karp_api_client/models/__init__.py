"""Models used by Karp API."""

from .entry_dto import EntryDto
from .entry_dto_entry import EntryDtoEntry
from .query_response import QueryResponse
from .validation_error import ValidationError

__all__ = ["EntryDto", "EntryDtoEntry", "QueryResponse", "ValidationError"]
