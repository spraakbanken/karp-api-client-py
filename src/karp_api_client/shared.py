"""Utility types."""

from collections.abc import MutableMapping
from http import HTTPStatus
from typing import Generic, Literal, Optional, TypeVar

import attrs


class Unset:
    def __bool__(self) -> Literal[False]:
        return False

    def __str__(self) -> str:
        return "UNSET"

    def __repr__(self) -> str:
        return "UNSET"


UNSET: Unset = Unset()
T = TypeVar("T")


@attrs.define
class Response(Generic[T]):
    """A response from an endpoint."""

    status_code: HTTPStatus
    content: bytes
    headers: MutableMapping[str, str]
    parsed: Optional[T]


__all__ = ["Response"]
