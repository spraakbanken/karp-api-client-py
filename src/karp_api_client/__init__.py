"""A client library for accessing Karp API."""

from karp_api_client.client import AuthenticatedClient, Client

__all__ = (
    "AuthenticatedClient",
    "Client",
)
