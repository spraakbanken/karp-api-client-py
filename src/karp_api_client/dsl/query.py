"""Karp Query DSL."""

import copy
from collections.abc import Callable
from typing import Any, ClassVar, Union

from karp_api_client.shared import UNSET, Unset

try:
    from typing import Self  # type: ignore [attr-defined]
except ImportError:
    from typing_extensions import Self


class Query:
    """Base class for all queries."""

    _param_defs: ClassVar[dict[str, dict[str, Union[str, bool]]]] = {}

    def __init__(self, **params: Any) -> None:
        """Construct a Query."""
        self._params: dict[str, Any] = {}
        for pname, pvalue in params.items():
            self._params[pname] = pvalue

    # Add type annotations for methods not defined in every subclass
    __ror__: ClassVar[Callable[["Query", "Query"], "Query"]]

    def __or__(self, other: "Query") -> "Query":
        """Combine queries with or."""
        # make sure we give queries that know how to combine themselves
        # preference
        if hasattr(other, "__ror__"):
            return other.__ror__(self)
        return Or(self, other)

    def _clone(self) -> Self:
        c = self.__class__()
        for attr in self._params:
            c._params[attr] = copy.copy(self._params[attr])
        return c

    def __getattr__(self, name: str) -> Any:
        """Read a value if it exists."""
        value = None
        try:
            value = self._params[name]
        except KeyError:
            if (pinfo := self._param_defs.get(name)) and pinfo.get("multi"):
                value = self._params.setdefault(name, [])
        if value is None:
            raise AttributeError(f"{self.__class__.__name__!r} has no attribute {name!r}")
        return value


class Equals(Query):
    """Find all entries where the `field` equals `value`.

    Stricter than `contains`.
    """

    _param_defs: ClassVar[dict[str, dict[str, Union[str, bool]]]] = {
        "field": {"type": "query", "multi": False},
        "value": {"type": "query", "multi": False},
    }

    def __init__(self, _field: Union[str, Unset] = UNSET, _value: Union[str, Unset] = UNSET, **kwargs: Any) -> None:
        """Constuct a Equals query."""
        if _field is not UNSET:
            kwargs[str(_field)] = _value
        super().__init__(**kwargs)

    def __str__(self) -> str:
        """Format this query as the Karp Api."""
        return f"equals|{self.field}|{self.value}"


class Or(Query):
    """Find all entries that matches any of the queries."""

    _param_defs: ClassVar[dict[str, dict[str, Union[str, bool]]]] = {"ors": {"type": "query", "multi": True}}

    def __init__(self, *queries: Query) -> None:
        """Construct an Or query by combining two queries."""
        super().__init__(ors=list(queries))

    def __or__(self, other: "Query") -> "Query":
        """Combine other query with or."""
        q = self._clone()
        if isinstance(other, Or):
            q.ors.extend(other.ors)
        else:
            q_other = other._clone()
            q.ors.append(q_other)
        return q

    def __str__(self) -> str:
        """Format for Karp API."""
        queries = "||".join(str(q) for q in self.ors)
        return f"or({queries})"
