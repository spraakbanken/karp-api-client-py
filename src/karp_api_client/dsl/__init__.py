
from .query import Equals, Query

__all__ = ["Equals", "Query"]
