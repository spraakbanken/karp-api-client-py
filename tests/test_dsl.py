from karp_api_client.dsl import Equals, Or


def test_dsl(snapshot) -> None:  # noqa: ANN001
    q = Equals(field="baseform", value="agha") | Equals(field="baseform", value="agin")

    actual = str(q)

    assert actual == snapshot


def test_empty_or() -> None:
    q = Or()

    assert q.ors == []


def test_equals_combined(snapshot) -> None:  # noqa: ANN001
    q = Or()

    q |= Equals(field="baseform", value="agha")
    q |= Equals(field="baseform", value="agin")

    actual = str(q)

    assert actual == snapshot
