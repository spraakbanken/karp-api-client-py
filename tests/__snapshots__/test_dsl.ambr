# serializer version: 1
# name: test_dsl
  'or(equals|baseform|agha||equals|baseform|agin)'
# ---
# name: test_equals_combined
  'or(equals|baseform|agha||equals|baseform|agin)'
# ---
